src/censored.jl

```diff
@@ -0,0 +1,234 @@
+`Inf`. The function throws an error if `l > u`.
+
+The function falls back to constructing a [`Censored`](@ref) wrapper.
+
+# Implementation
+
+To implement a specialized censored form for distributions of type `D`, the method
+`censored(d::D, l::T, u::T) where {T <: Real}` should be implemented.
+"""
+function censored(d::UnivariateDistribution, l::Real, u::Real)
+    return censored(d, promote(l, u)...)
+end
+censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real} = Censored(d, l, u)
+
+"""
+    Censored
+
+Generic wrapper for a censored distribution.
+"""
+struct Censored{D<:UnivariateDistribution, S<:ValueSupport, T <: Real} <: UnivariateDistribution{S}
+    uncensored::D      # the original distribution (uncensored)
+    lower::T      # lower bound
+    upper::T      # upper bound
+    function Censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real}
+        l ≤ u || error("the lower bound must be less than or equal to the upper bound")
+        new{typeof(d), value_support(typeof(d)), T}(d, l, u)
+    end
+end
+
+params(d::Censored) = tuple(params(d.uncensored)..., d.lower, d.upper)
+partype(d::Censored) = partype(d.uncensored)
+Base.eltype(::Type{<:Censored{D}}) where {D} = eltype(D)
+
+#### Range and Support
+
+islowerbounded(d::Censored) = islowerbounded(d.uncensored) || isfinite(d.lower)
+isupperbounded(d::Censored) = isupperbounded(d.uncensored) || isfinite(d.upper)
+
+minimum(d::Censored) = max(minimum(d.uncensored), d.lower)
+maximum(d::Censored) = min(maximum(d.uncensored), d.upper)
+
+function insupport(d::Censored{<:UnivariateDistribution}, x::Real)
+    return d.lower ≤ x ≤ d.upper && insupport(d.uncensored, x)
+end
+
+#### Show
+
+function show(io::IO, d::Censored)
+    print(io, "Censored(")
+    d0 = d.uncensored
+    uml, namevals = _use_multline_show(d0)
+    uml ? show_multline(io, d0, namevals; newline=false) : show_oneline(io, d0, namevals)
+    print(io, ", range=(", d.lower, ", ", d.upper, "))")
+end
+
+_use_multline_show(d::Censored) = _use_multline_show(d.uncensored)
+
+
+#### Statistics
+
+quantile(d::Censored, p::Real) = clamp(quantile(d.uncensored, p), d.lower, d.upper)
+
+median(d::Censored) = clamp(median(d.uncensored), d.lower, d.upper)
+
+function mean(d::Censored)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    dtrunc = truncated(d.uncensored, lower, upper)
+    prob_lower = if value_support(typeof(d)) === Discrete && isfinite(lower)
+        cdf(d0, lower) - pdf(d0, lower)
+    else
+        cdf(d0, lower)
+    end
+    prob_upper = ccdf(d0, upper)
+    prob_interval = 1 - (prob_lower + prob_upper)
+
+    μ = prob_interval * mean(dtrunc)
+    if isfinite(lower)
+        μ += prob_lower * lower
+    end
+    if isfinite(upper)
+        μ += prob_upper * upper
+    end
+    return μ
+end
+
+function var(d::Censored)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    dtrunc = truncated(d.uncensored, lower, upper)
```

Comment:
Elegant!

Reply:
All expectations take a similar form (same as the mixture model representation); this just fuses the two expectations. I haven't checked yet if we can do less work by unrolling a [weighted Welford algorithm](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Weighted_incremental_algorithm).

Reply:
Welford didn't seem to be the right algorithm. I used the simplification for mixture distributions in https://en.wikipedia.org/wiki/Mixture_distribution#Moments, but it replaces 2 additions with 3 multiplications, so this version is faster.